### ISSUE TITLE
Nested model not training for distributed gradient tapes

### ISSUE THREAD

**Issue by Orpheus23** [tensorflow/tensorflow#48361]
**System information**
- Have I written custom code: Yes
- OS Platform and Distribution: Ubuntu18.04
- TensorFlow installed from: pip
- TensorFlow version: 2.4.1
- Python version: 3.7.1
- CUDA/cuDNN version: 11.2
- GPU model and memory: Tesla V100 each with 16 GB

**Describe the current behavior**
Nested model (Encoder) not training/ model parameters not changing

**Describe the expected behavior**
Nested (Encoder) model should be trained

I had created a model for an encoder that is used in both the generator as well as the discriminator. While the same encoder model shows in total params for both the generator and discriminator, it doesn't train at all. When the same encoder model is later called on random inputs it will always give the same result of 0.

    
    tf.keras.backend.clear_session()

    gamma_init = tf.keras.initializers.RandomNormal(mean=0.02, stddev=0.02) 
    def ENCODER(shape=(int(456/2), int(456/2), 3)):
        y0 = tf.keras.Input(shape=shape,name = 'encoder')
        y1 = tf.keras.applications.MobileNetV3Small(include_top=False,weights=None,input_shape=shape)(y0)
        y1 = tf.keras.layers.GlobalAveragePooling2D()(y1)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(y1)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        encoding = tf.keras.layers.Dense(32)(x)
        encoding = tf.keras.layers.Lambda(lambda x: tf.math.l2_normalize(x,axis=1),name="L2_normalized_encodings")(encoding)
        model = tf.keras.Model(y0, encoding)
        return model
    
    def EMBEDDER(shape = (1,),classes = NUM_CLASSES):
        input_ = tf.keras.Input(shape=shape,name = 'embedder')
        embedings_raw = tf.keras.layers.Embedding(NUM_CLASSES,32)(input_)
        embedings = tf.keras.layers.Lambda(lambda x:x[:,0])(embedings_raw)
        embedings = tf.keras.layers.Lambda(lambda x: tf.math.l2_normalize(x,axis=1),name="L2_normalized_embeddings")(embedings)
        return tf.keras.Model(input_,embedings)
    
    def DISCRIMANTOR(encoder):
        generation = tf.keras.Input(shape=(int(456/2), int(456/2), 3), name='input_1_disc')
        embedings = tf.keras.Input(shape=(32,), name='input_2_disc')
        delta_layer = encoder(generation,training=True)
        ##SIGMOID OUTPUT MODULE
        modified_delta = tf.keras.layers.Add()([delta_layer,embedings])
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(modified_delta)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        x = tf.keras.layers.Dense(16)(x)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(x)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        x = tf.keras.layers.Dense(8)(x)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(x)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        output_layer = tf.keras.layers.Dense(1,activation = 'sigmoid')(x)
        return tf.keras.Model(inputs =[generation,embedings],outputs = [output_layer])

    def GENERATOR():
        encodings = tf.keras.Input(shape=(32,), name='input_1_gen')
        embedings = tf.keras.Input(shape=(32,), name='input_2_gen')
        composed_layer = tf.keras.layers.Concatenate(axis=-1)([encodings,embedings])
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(composed_layer)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        x = tf.keras.layers.Dense(256)(x)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(x)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        encodings_for_label = tf.keras.layers.Dense(128*2*2)(x)
        image_format = tf.keras.layers.Reshape((2,2, 128), name='de_reshape')(encodings_for_label)
        first_image = tf.keras.layers.Conv2DTranspose(filters = 256,kernel_size=(2, 2) ,strides = 2)(image_format)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(first_image)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        second_image = tf.keras.layers.Conv2DTranspose(filters = 256,kernel_size=(3, 3),strides = 2)(x)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(second_image)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        third_image = tf.keras.layers.Conv2DTranspose(filters = 128,kernel_size=(3, 3) ,strides = 2)(x)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(third_image)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        fourth_image = tf.keras.layers.Conv2DTranspose(filters = 128,kernel_size=(1, 1))(tf.keras.layers.Add()([x,tf.keras.layers.Conv2DTranspose(filters = 128,kernel_size=(7, 7),strides = 4)(first_image)]))
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(fourth_image)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        second_image = tf.keras.layers.Conv2DTranspose(filters = 64,kernel_size=(2, 2),strides = 3)(x)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(second_image)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        third_image = tf.keras.layers.Conv2DTranspose(filters = 64,kernel_size=(2, 2) ,strides = 2)(x)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(third_image)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        
        fourth_image =  tf.keras.layers.Conv2DTranspose(filters = 32,kernel_size=(1, 1))(tf.keras.layers.Add()([x,tf.keras.layers.Conv2DTranspose(filters = 64,kernel_size=(6, 6),strides = 6)(fourth_image)]))
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(fourth_image)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        second_image = tf.keras.layers.Conv2DTranspose(filters = 16,kernel_size=(1, 1),strides = 2)(x)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(second_image)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        third_image = tf.keras.layers.Conv2DTranspose(filters = 8,kernel_size=(1, 1))(x)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(third_image)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        fourth_image = tf.keras.layers.Conv2DTranspose(filters = 3,kernel_size = 1,name="fourth_output")(x)
        return tf.keras.Model(inputs=[encodings,embedings],outputs = [fourth_image]) def DISCRIMANTOR(encoder):
        generation = tf.keras.Input(shape=(int(456/2), int(456/2), 3), name='input_1_disc')
        embedings = tf.keras.Input(shape=(32,), name='input_2_disc')
        delta_layer = encoder(generation,training=True)
        ##SIGMOID OUTPUT MODULE
        modified_delta = tf.keras.layers.Add()([delta_layer,embedings])
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(modified_delta)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        x = tf.keras.layers.Dense(16)(x)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(x)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        x = tf.keras.layers.Dense(8)(x)
        x = tfa.layers.InstanceNormalization(gamma_initializer=gamma_init)(x)
        x = tf.keras.layers.Dropout(0.4)(x)
        x = tf.keras.layers.LeakyReLU(alpha=0.3)(x)
        output_layer = tf.keras.layers.Dense(1,activation = 'sigmoid')(x)
        return tf.keras.Model(inputs =[generation,embedings],outputs = [output_layer])
```


for the training, I use the following loop:- 

with stratergy.scope()
    input_image =tf.keras.Input(shape=(228,228, 3))
    input_label =tf.keras.Input(shape=(1,))
    ENC = ENCODER()
    encoding = ENC(input_image)
    embeding = EMBEDDER()(input_label)
    image_new = GENERATOR()([encoding,embeding])
    reality_check = DISCRIMANTOR(ENC)([input_image,embeding])
    discriminator = tf.keras.Model(inputs = [input_image,input_label],outputs = [reality_check])   
    gen_trainer = tf.keras.Model(inputs = [input_image,input_label],outputs = [encoding,image_new])

def train_complete():
    def train_step(inputs):
        image, labels = inputs
        #Precitions for generator:- [image,encoding_original,encoding_new]
        #Precitions for discriminator:- [softmax output,embedding+centre]
        #DISCRIMINATOR TRAINING
        discrim_labels = tf.concat([tf.ones_like(labels),tf.zeros_like(labels)],axis = 0)
        gen_discim = tf.concat([tf.zeros_like(labels),tf.ones_like(labels)],axis = 0)
        with tf.GradientTape(persistent=True) as tape:
            image_,encoding_original,encoding_new = gen_trainer(inputs,training=True)
            image_loss = GEN_IMAGE_LOSS(image,image_)
            encoding_loss = ENCODER_LOSS(encoding_original,encoding_new)
            
            binary_preds = discriminator([tf.concat([image/1.,image_],axis = 0),tf.concat([labels,labels],axis = 0)],training=True)
            binary_loss = DISC_SOFTMAX_LOSS(discrim_labels,binary_preds)
            loss_final_gen = encoding_loss + 10*image_loss + DISC_SOFTMAX_LOSS(gen_discim,binary_preds)
            loss_final_disc = binary_loss + centre_trip_loss
            
        gradients = tape.gradient(loss_final_disc, discriminator.trainable_variables)
        
        var_list_disc = discriminator.trainable_variables
        
        #CLIP BY VALUE
        sam_gradients = [tf.clip_by_value(grad, -1., 1.) for grad in gradients]
        #APPLY TO GRADIENTS
        discriminator.optimizer.apply_gradients(zip(sam_gradients, discriminator.trainable_variables),experimental_aggregate_gradients=False)
        
        
        ##GENERATOR TRAINING
          
        gradients = tape.gradient(loss_final_gen, gen_trainer.trainable_variables)
        
        var_list_gen = gen_trainer.trainable_variables
        
        #CLIP BY VALUE
        sam_gradients = [tf.clip_by_value(grad, -1., 1.) for grad in gradients]
        #APPLY TO GRADIENTS
        gen_trainer.optimizer.apply_gradients(zip(sam_gradients, gen_trainer.trainable_variables),experimental_aggregate_gradients=False)
        del tape
        return binary_loss,image_loss,encoding_loss
    @tf.function
    def distributed_train_step(inputs):
        loss1,loss2,loss3= strategy.run(train_step, args=(inputs,))
        #train_metric.update_state(strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_losses,axis=None))
        loss1 = strategy.reduce(tf.distribute.ReduceOp.SUM, loss1,axis=None)
        loss2 = strategy.reduce(tf.distribute.ReduceOp.SUM, loss2,axis=None)
        loss3 = strategy.reduce(tf.distribute.ReduceOp.SUM, loss3,axis=None)
        return loss1,loss2,loss3
    return distributed_train_step
train_function = train_complete()
for _ in range(80):
    total_loss = np.array([0.0,0.0,0.0])
    num_batches = 0
    for x in train_dataset:
          total_loss += train_function(x)
          num_batches += 1
          train_loss = total_loss / num_batches
```

The loss functions here are as follows:-

```
    @tf.function
    def DISC_EMBED_LOSS(labels, predictions):
        per_example_loss = TripletSemiHardLoss(distance_metric = pairwise_distance,reduction=tf.keras.losses.Reduction.NONE)(labels, predictions)
        return per_example_loss/ strategy.num_replicas_in_sync
    @tf.function
    def ENCODER_LOSS(labels,predictions):
        per_example_loss = tf.nn.compute_average_loss(1.+tf.keras.losses.CosineSimilarity(axis=-1,reduction=tf.keras.losses.Reduction.NONE)(labels,predictions), global_batch_size=GLOBAL_BATCH_SIZE)
        return per_example_loss
    @tf.function
    def DISC_SOFTMAX_LOSS(labels, predictions):
        per_example_loss =  tf.keras.losses.BinaryCrossentropy(reduction=tf.keras.losses.Reduction.NONE,from_logits=True)(labels, predictions)
        return per_example_loss/ strategy.num_replicas_in_sync
    @tf.function
    def GEN_IMAGE_LOSS(labels, predictions):
        per_example_loss = tf.math.reduce_mean(tf.math.abs(labels/255. - predictions/255.))
        return per_example_loss/ strategy.num_replicas_in_sync
```

With all this being as it is when I call the encoder on any other value the prediction is always 0 irrespective of input or number of epochs. Which I believe is due to it not getting trained. What could be the issue? Does the error lie with gradient tapes, because when the encoder was trained by itself with train_on_batch, the encoder's weights were updated. Also, note there isn't any graph disconnection error in the generator and discriminator.

**Comment by nikitamaia** [comment]
Hi @Orpheus23, a few questions:

- what strategy are you using? (I'm assuming `MirroredStrategy`?)
- Are you distributing your data with [experimental_distribute_dataset](https://www.tensorflow.org/api_docs/python/tf/distribute/Strategy#experimental_distribute_dataset)?
- have you verified that things are working as expected in the non distributed case?

**Comment by Orpheus23** [comment]
Hello, thanks for the reply,

- I am using MirroredStrategy.
- Yes the data is distributed with experimental_distribute_dataset.
- Things aren't working as expected even in the non-distributed case, the same thing is happening there too.

**Comment by nikitamaia** [comment]
Ah okay, in that case I think it makes sense to try and debug your model first without a distribution strategy and see if you can get it to train. In general, you want to make sure your model trains as expected in the non distributed case before modifying the code for MirroredStrategy.

**Comment by Orpheus23** [comment]
Oh okay well, my original issue was the model, not training. I solved it by creating my model within the scope rather in one shot rather than by creating it via the functions. Anyhow, thanks for the help.
